modules/httpcache/httpcache.go

```diff
@@ -40,20 +43,45 @@ func HandleTimeCache(req *http.Request, w http.ResponseWriter, fi os.FileInfo) (
 		}
 	}
 
-	w.Header().Set("Cache-Control", GetCacheControl())
 	w.Header().Set("Last-Modified", fi.ModTime().Format(http.TimeFormat))
 	return false
 }
 
-// HandleEtagCache handles ETag-based caching for a HTTP request
-func HandleEtagCache(req *http.Request, w http.ResponseWriter, fi os.FileInfo) (handled bool) {
+// HandleFileEtagCache handles ETag-based caching for a HTTP request
+func HandleFileEtagCache(req *http.Request, w http.ResponseWriter, fi os.FileInfo) (handled bool) {
 	etag := generateETag(fi)
-	if req.Header.Get("If-None-Match") == etag {
-		w.WriteHeader(http.StatusNotModified)
-		return true
-	}
+	return HandleGenericETagCache(req, w, etag)
+}
 
+// HandleGenericETagCache handles ETag-based caching for a HTTP request.
+// It returns true if the request was handled.
+func HandleGenericETagCache(req *http.Request, w http.ResponseWriter, etag string) (handled bool) {
+	if len(etag) > 0 {
+		w.Header().Set("Etag", etag)
+		if checkIfNoneMatchIsValid(req, etag) {
+			w.WriteHeader(http.StatusNotModified)
+			return true
+		}
+	}
 	w.Header().Set("Cache-Control", GetCacheControl())
-	w.Header().Set("ETag", etag)
+	return false
+}
+
+// checkIfNoneMatchIsValid tests if the header If-None-Match matches the ETag
+func checkIfNoneMatchIsValid(req *http.Request, etag string) bool {
+	ifNoneMatch := req.Header.Get("If-None-Match")
+	if len(ifNoneMatch) > 0 {
+		if ifNoneMatch == "*" {
+			return true
+		}
```

Comment:
Could also opt to remove this check. In the common case (which should be optimized), it will likely never appear and even if it does appear, it would still work fine with the checks below, unless we start sending `Etag: *` which we would never do. Let's remove it for the sake of performance.

Reply:
`*` gets send by the client not the server.
> If the field-value is "*", the condition is false if the origin server has a current representation for the target resource

Reply:
Yeah sorry, I meant `If-None-Match: *` from the client, which is probably not something we will see in the common case, so I thought it needs not to  be optimized.

Reply:
Personally I'm not a fan of this micro-optimization but I have removed both. It still works :wink:

Reply:
Thanks, should always optimize for the most likely cases 😉